src/Resources/ResourceManager/SdkModels/Deployments/DeploymentOperationErrorInfo.cs

```diff
@@ -0,0 +1,142 @@
+                }
+                
+                ErrorMessages.Add(sb.ToString());
+            }            
+        }
+
+        public string GetErrorMessagesWithOperationId(string deploymentName = null)
+        {
+            if (ErrorMessages.Count == 0)
+                return String.Empty;
+
+            var sb = new StringBuilder();
+
+            int maxErrors = ErrorMessages.Count > MaxErrorsToShow
+               ? MaxErrorsToShow
+               : ErrorMessages.Count;
+
+            sb.AppendFormat(ProjectResources.DeploymentOperationOuterError, deploymentName, maxErrors, ErrorMessages.Count);
+            sb.Append(string.Join(Environment.NewLine, ErrorMessages.Take(maxErrors)));
+            sb.AppendLine().AppendFormat(ProjectResources.DeploymentOperationId, this.RequestId);
+            
+            return sb.ToString();
+        }
+
+
+        public void SetRequestIdFromResponseHeaders(HttpResponseMessage response)
+        {
+            RequestId = string.Empty;
+
+            if (response?.Headers != null && response.Headers.TryGetValues("x-ms-request-id", out IEnumerable<string> requestIdValues))
+            {
+                RequestId = string.Join(";", requestIdValues);
+            }
+        }
+        #endregion
+
+        #region Static Methods
+        public static string GetErrorMessageWithDetails(ErrorResponse error)
+        {
+            if (error == null) return null;
+
+            if (error.Details == null)
+            {
+                return string.Format(ProjectResources.DeploymentOperationErrorMessageNoDetails, error.Code, error.Message);
+            }
+
+            string errorDetail = null;
+
+            foreach (ErrorResponse detail in error.Details)
+            {
+                errorDetail += GetErrorMessageWithDetails(detail);
+            }
+
+            return string.Format(ProjectResources.DeploymentOperationErrorMessage, error.Code, error.Message, errorDetail);            
+        }
+
+        public static ErrorResponse DeserializeDeploymentOperationError(string statusMessage)
+        {
+            if (statusMessage == null)
+            {
+                return null;
+            }
+
+            ErrorResponse error = null;
+            dynamic dynamicStatusMessage;
+
+            try
+            {
+                dynamicStatusMessage = JsonConvert.DeserializeObject(statusMessage);
+                error = JsonConvert.DeserializeObject<ErrorResponse>(dynamicStatusMessage?.error?.ToString());
+            }
+            catch
+            {
+                // We'll ignore the exception if we can't properly deserialize the JSON.
+                // The reason for that can be 1- statusMessage is not a valid JSON(it can be a string when it's the outer generic error message)
+                // which will result is DeserializeObject exception. 2- If error is not properly modeled as ErrorResponse.
+            }
+
+            return error;
```

Comment:
This will return null if you get an exception, which will crash at += with NullReferenceException in the recursive call.

Reply:
The recursive call `GetErrorMessageWithDetails` has a check for null condition, it shouldn't be able to go into the += part of the code.

Reply:
Cool, just need to be sure every caller does the check. Since the null only happens on an error path, it's a slightly non-obvious return value. Not a big deal, but it might be safer to use empty string.